recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,196 @@
+
+```
+...
+ recipeConfig: {
+    terraform: {
+      ...
+      providers: {
+        [azurerm: {  
+          subscriptionid: 1234
+          secrets: {
+            source: secretStoreAz.id
+          }
+        },
+        azurerm: {
+        subscriptionid: 1234
+        alias: 'az-paymentservice'
+        secrets: {
+          source: secretStoreAzPayment.id
+        }         
+      }]
+    } 
+...       
+```
+Configuration for providers as described in this document will take precedence over provider credentials stored in UCP (currently these include azurerm, aws, kubernetes providers). So, for eg. In the scenario where credentials for Azure are saved with UCP during Radius install and a Terraform recipe created by an operator declares 'azurerm' under the the *required_providers* block; If there exists a provider configuration under the *providers* block under *recipeConfig*, these would take precedence and used to build the Terraform configuration file instead of Azure credentials stored in UCP. 
+
+
+Example :
+``` diff
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
+  recipeConfig: {
+    terraform: {
+      authentication:{
+        ...        
+      }
++      providers: {
++        [azurerm: {  
++          subscriptionid: 1234
++          secrets: {
++            source: secretStoreAz.id
++          }
++        },
++        azurerm: {
++         subscriptionid: 1234
++         alias: 'az-paymentservice'
++         secrets: {
++           source: secretStoreAzPayment.id
++         }         
++       },
++       gcp: {
++         project: 1234
++         regions: ['us-east1', 'us-west1']
++       },
++       oraclepass: {
++         database_endpoint = "..."
++         java_endpoint     = "..."
++         mysql_endpoint    = "..."
++       }]
++     }
++     env: {
++       key: value
++       secrets: {
++        source: secretStoreOCICred.id
++      }
++     }
+    }
+  }
+    recipes: {      
+      ...
+    }
+  }
+}
+```
+
+
+Limitations: Customers may store sensitive data in other formats which may not be supported. eg. sensitive data is stored on files, which customers will not currently be able to load on disk in applications-rp where tf init/apply commands are run.
+Containerization work may alleviate this limitations. Further design work will be needed for towards this which is planned for the near future. 
+
+### API design (if applicable) (TBD)
+
+TBD
+
+## Alternatives considered
+
+Mentioned under Limitations above, work to containerize running terraform jobs was considered as a precursor to this work. The time sensitivity for unblocking customers on ability to configure and use providers they use today was given a priority and current design held. Containerization will be taken up as a parallel effort in the near future.
+
```

Comment:
Do we need a way to specific configuration with specific recipes?

Reply:
As discussed, this was considered and we've went ahead with first targetting setting up configuration at environment level before getting more granular at the individual recipe level.